core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -382,7 +382,10 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
     if (newProducerId < 0)
       throw new IllegalArgumentException(s"Illegal new producer id $newProducerId")
 
-    if (newEpoch < 0)
+    // The epoch is initialized to NO_PRODUCER_EPOCH when the TransactionMetadata
+    // is created for the first time and it could stay like this until transitioning
+    // to Dead.
+    if (newState != Dead && newEpoch < 0)
```

Comment:
To confirm -- the issue only arises when we transition to Dead + try to expire? 
We won't see any other transitions until the epoch is written correctly?

Reply:
Right. We basically transitions from Empty (without a producer epoch) to Dead. My thinking is that we should basically always accept to transition to Dead.

Reply:
Makes sense to me. Thanks. Just wanted to make sure we couldn't transition to another state (in the current code) and we would hit this error.